web/modules/custom/usergreeting/src/GreetingTime.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\usergreeting;
+
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+
+/**
+ * Provides a greeting service for users dependent on the time of  day.
+ */
+final class GreetingTime {
+
+  use StringTranslationTrait;
+
+  /**
+   * The user account.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $account;
+
+  /**
+   * The time output.
+   *
+   * @var \Drupal\Component\Datetime\TimeInterface
+   */
+  protected $timeOutput;
+
+  /**
+   * GreetingTime constructor.
+   *
+   * @param \Drupal\Component\Datetime\TimeInterface $timeOutput
+   *   The timestamp output.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The user account.
+   */
+  public function __construct(TimeInterface $timeOutput, AccountInterface $account) {
+    $this->timeOutput = $timeOutput;
+    $this->account = $account;
+  }
+
+  /**
+   * Returns the greeting.
+   *
+   * @return string
+   *   The greeting message output.
+   */
+  public function greetingMessage(): string {
+    $time_output = (int) date('G', $this->timeOutput->getRequestTime());
```

Comment:
I checked the code, logged in with a user, changed my timezone to Vancouver and it's still God evening. :) You still need to use the current user's timezone and format it so the right time is calculated.


This can probably help: 
https://api.drupal.org/api/drupal/core%21modules%21user%21src%21Entity%21User.php/function/User%3A%3AgetTimeZone/8.2.x
https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Datetime%21DateFormatter.php/function/DateFormatter%3A%3Aformat/8.2.x

Reply:
Berci told me you checked this together since then, it seems to be more of a caching issue and not a date issue then.